source/slang/slang-lower-to-ir.cpp

```diff
@@ -6612,9 +6650,94 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
         IRGeneric*  parentGeneric)
     {
         IRInst* v = val;
+
+        IRInst* returnType = v->getFullType();
+
         while (parentGeneric)
         {
+            // Create a universal type in `outterBlock` that will be used
+            // as the type of this generic inst. The return value of the
+            // generic inst will have a specialized type.
+            // For example, if we have a generic function
+            // g0 = generic<T> { return f: T->int }
+            // The type for `g0` should be:
+            // g0Type = generic<T1> { return IRFuncType{T1->int} }
```

Comment:
Doesn't this lead to infinite regress, since now `g0Type` needs a type? I guess at that point we have:
```
g0TypeType = generic<T2> { return IRType; }
```
And then `g0TypeType` can be used as the "type" of all generics with a single type parameter that return a type (that is, for the kind `Type -> Type`). We'd have to make sure to actually re-use those types.

That's definitely the bit where I'm still concerned that having "the type of a generic is a generic" isn't 100% right. We really need something that can represent `forall T . T->int` as a *type* and not a specific generic *value*.

This is getting to be a worrying hole in our IR as we start to care more and more about getting the types of generics "right."

Reply:
Note: The challenge for making `forall` work as a type in our IR is basically that we've wanted types (other than nominal types) to be defined 100% by their operands (no need to inspect decorations/children). To support a type like `forall T . T->int` that means that `T->int` needs to somehow be an *operand* of the `forall` instruction, but that means that the declaration of `T` cannot somehow be a nested `IRParam` scoped under the `forall`.

I *think* the solution in this case is an encoding with something a bit like De Brujin indices:

* We have an IR instruction like `placeholderParam(index) : Type` which is uniquely defined by the `index` and `Type`.
* A `forall T0, T1, ... . whatever` instruction is encoded as `univeralQuantify(whatever, T0, T1, ...)` where each of the `Ti` is a `placeholderParam` that is suitable to the kind of the generic parameter and its index `i`. E.g., the `T0` might be `placeholderParam(0) : Type` for the generic in this example.
* The `whatever` result of a `universalQuantify` should then be an ordinary IR instruction (usually a type), but one that may make use of the `placeholderParam`s from whatever quantifies it. In the example above, that would be `IRFuncType(placeHolderParam(0)->int)`

Thus for the example, we might have:

```
let T1 = placeholderParam(0) : Type;
let F = IRFuncType(T1 -> int);
let g0Type = universalQuantify(T1 -> F);
let g0 : g0Type = generic<...> { ... } // unchanged from current encoding
```

This approach avoids using IR generics to encode the types of things, and seems like it is logically more consistent.